lib/restHandlers.js

```diff
@@ -0,0 +1,226 @@
+(function() {
+	'use strict';
+
+	var _ = require('lodash'),
+		qs = require('qs'),
+		apiUrl = require(__dirname + '/apiUrl'),
+		log = require(__dirname + '/log'),
+		rest = require(__dirname + '/rest');
+
+	function RestHandlers(options) {
+		var defaults = {
+			apiToken: null
+		};
+
+		this.options = _.extend({}, defaults, options || {});
+
+		return this;
+	}
+
+	RestHandlers.prototype.genericResponse = function(method, object, responseBody, callback, rawRequest, rawResponse) {
+
+		if (_.isString(responseBody)) {
+			try {
+				responseBody = JSON.parse(responseBody);
+			}
+			catch (err) {}
+		}
+
+		log('Handling response of ' + method + ' ' + object + ': ' + rawRequest.url.href + (rawRequest.url.href.indexOf(rawRequest.url.query) === -1 ? '?' + rawRequest.url.query : ''));
+
+		if (responseBody.success === true) {
+			if (_.isFunction(callback)) {
+				callback(null, responseBody.data, responseBody.additional_data, rawRequest, rawResponse);
+			}
+		}
+		else {
+			if (rawResponse.statusCode && (rawResponse.statusCode.toString().substr(0,1) == '4' || rawResponse.statusCode.toString().substr(0,1) == '5')) {
+
+				var errorObject = new Error();
+
+				if (_.isObject(rawResponse._error)) {
+					errorObject = rawResponse._error;
+				}
+				else {
+					errorObject = new Error('Got HTTP response ' + rawResponse.statusCode + ' from Pipedrive API');
+
+					if (responseBody.error) {
+						errorObject = new Error('Pipedrive API error:' + responseBody.error);
+					}
+					else {
+						try {
+							var errorBody = JSON.parse(rawResponse.rawEncoded.toString('utf8'));
+							if (!_.isUndefined(errorBody.error)) {
+								errorObject = new Error('Pipedrive API error:' + errorBody.error);
+							}
+						}
+						catch (err) {}
+					}
+				}
+
+				if (_.isFunction(callback)) {
+					callback(errorObject, null, null, rawRequest, rawResponse);
+				}
+			}
+			else {
+				if (_.isFunction(callback)) {
+					callback(null, responseBody.data || {}, responseBody.additional_data || {}, rawRequest, rawResponse);
+				}
+			}
+		}
+
+	};
+
+	// GET /items
+	RestHandlers.prototype.listItems = function(object, params, callback) {
+		var that = this,
+			paramsToSupply = _.extend({}, _.isObject(params) ? params : {}, this.options.apiToken ? { api_token: this.options.apiToken } : {}),
+			dataObject = object == 'authorizations' ? { multipart: false, data: paramsToSupply } : { query: qs.stringify(paramsToSupply) },
```

Comment:
here and in the rest of the file, we should probably use "self" instead of "that" for consistency